File: app/src/main/java/ir/ghararemaghzha/game/fragments/HighscoreFragment.java
package ir.ghararemaghzha.game.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textview.MaterialTextView;

import java.util.List;

import ir.ghararemaghzha.game.R;
import ir.ghararemaghzha.game.adapters.HighscoreAdapter;
import ir.ghararemaghzha.game.classes.MySharedPreference;
import ir.ghararemaghzha.game.classes.Utils;
import ir.ghararemaghzha.game.data.RetrofitClient;
import ir.ghararemaghzha.game.models.HighscoreModel;
import ir.ghararemaghzha.game.models.HighscoreResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class HighscoreFragment extends Fragment {
    private Context context;
    private FragmentActivity activity;
    private ConstraintLayout loading;
    private RecyclerView recyclerView;
    private HighscoreAdapter adapter;

    public HighscoreFragment() {
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_highscore, container, false);
        context = getContext();
        activity = getActivity();
        init(v);

        return v;
    }

    private void init(View v) {
        ((MaterialTextView)activity.findViewById(R.id.toolbar_title)).setText(R.string.highscore_title);

        recyclerView = v.findViewById(R.id.highscore_recycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(context));

        loading = v.findViewById(R.id.highscore_loading);

        getData();
    }


    private void getData() {
        String number = MySharedPreference.Companion.getInstance(context).getNumber();
        String token = MySharedPreference.Companion.getInstance(context).getAccessToken();
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(activity,true);
            return;
        }
        loading.setVisibility(View.VISIBLE);

        RetrofitClient.Companion.getInstance().getApi().getHighscoreList("Bearer " + token, number)
                .enqueue(new Callback<HighscoreResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<HighscoreResponse> call, @NonNull Response<HighscoreResponse> response) {
                        loading.setVisibility(View.GONE);
                        if (response.isSuccessful() && response.body() != null && response.body().getResult().equals("success")) {

                            boolean showUser=true;
                            List<HighscoreModel> data = response.body().getData();
                            for(int i=0;i<data.size();i++){
                                if (data.get(i).getUserId().equals(response.body().getUser().getUserId())) {
                                    showUser=false;
                                }
                            }

                            adapter = new HighscoreAdapter(activity,data,response.body().getUser(),showUser);
                            recyclerView.setAdapter(adapter);
                            adapter.notifyDataSetChanged();


                        } else if (response.code() == 401) {
                            Utils.logout(activity,true);
                        }else{
                            Utils.showInternetError(context, () -> getData());
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<HighscoreResponse> call, @NonNull Throwable t) {
                        loading.setVisibility(View.GONE);
                        Utils.showInternetError(context, () -> getData());

                    }
                });
    }
}

File: app/src/main/java/ir/ghararemaghzha/game/fragments/HighscoreFragment.kt
package ir.ghararemaghzha.game.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.textview.MaterialTextView
import ir.ghararemaghzha.game.R
import ir.ghararemaghzha.game.adapters.HighscoreAdapter
import ir.ghararemaghzha.game.classes.MySharedPreference
import ir.ghararemaghzha.game.classes.Utils
import ir.ghararemaghzha.game.classes.Utils.Companion.showInternetError
import ir.ghararemaghzha.game.data.Resource
import ir.ghararemaghzha.game.databinding.FragmentHighscoreBinding
import ir.ghararemaghzha.game.viewmodels.HighscoreViewModel

class HighscoreFragment : BaseFragment<HighscoreViewModel, FragmentHighscoreBinding>() {

    private lateinit var adapter: HighscoreAdapter
    private lateinit var number: String
    private lateinit var token: String

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        init()

        viewModel.highscoreResponse.observe(viewLifecycleOwner, { res ->
            when (res) {
                is Resource.Success -> {
                    if (res.value.result == "success") {
                        val data = res.value.data
                        val user = res.value.user
                        var showUser = true
                        for (i in data)
                            if (i.userId == user.userId) {
                                showUser = false
                                break
                            }

                        adapter = HighscoreAdapter(requireActivity(), data, user, showUser)
                        b.highscoreRecycler.adapter = adapter
                        b.highscoreLoading.visibility = View.GONE

                    } else {
                        showInternetError(requireContext()) { viewModel.getHighscoreList("Bearer $token", number) }
                        Toast.makeText(context, getString(R.string.general_error), Toast.LENGTH_SHORT).show()
                    }
                }

                is Resource.Failure -> {
                    if (res.isNetworkError) {
                        showInternetError(requireContext()) { viewModel.getHighscoreList("Bearer $token", number) }
                        Toast.makeText(context, getString(R.string.general_error), Toast.LENGTH_SHORT).show()
                    } else if (res.errorCode == 401) {
                        b.highscoreLoading.visibility = View.GONE
                        Utils.logout(requireActivity(), true)
                    }
                }
                is Resource.Loading -> {
                }
            }
        })
    }

    private fun init() {
        requireActivity().findViewById<MaterialTextView>(R.id.toolbar_title).setText(R.string.highscore_title)
        b.highscoreRecycler.layoutManager = LinearLayoutManager(context)
        b.highscoreLoading.visibility = View.VISIBLE

        number = MySharedPreference.getInstance(requireContext()).getNumber()
        token = MySharedPreference.getInstance(requireContext()).getAccessToken()
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(requireActivity(), true)
            return
        }

        viewModel.getHighscoreList("Bearer $token", number)
    }

    override fun getViewModel() = HighscoreViewModel::class.java
    override fun getFragmentBinding(inflater: LayoutInflater, container: ViewGroup?) = FragmentHighscoreBinding.inflate(inflater, container, false)
}